core/impl/src/main/java/com/blazebit/persistence/impl/AbstractCTECriteriaBuilder.java

```diff
@@ -201,38 +209,83 @@ public CTEInfo createCTEInfo() {
                 // so it is safe to just append the id to the path
-                Expression selectExpression = selectManager.getSelectInfos().get(bindingEntry.getValue()).getExpression();
+                Expression selectExpression = selectManager.getSelectInfos().get(tupleIndex).getExpression();
 
                 // TODO: Maybe also allow Treat, Case-When, Array?
                 if (selectExpression instanceof NullExpression) {
                     // When binding null, we don't have to adapt anything
                 } else if (selectExpression instanceof PathExpression) {
-                    PathExpression pathExpression = (PathExpression) selectExpression;
-                    // Only append the id if it's not already there
-                    if (!idAttribute.getName().equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
-                        pathExpression.getExpressions().add(new PropertyExpression(idAttribute.getName()));
+                    boolean firstBinding = true;
+                    final Collection<String> embeddedPropertyNames;
+
+                    if (lastAttribute.getPersistentAttributeType() == Attribute.PersistentAttributeType.EMBEDDED) {
+                        embeddedPropertyNames = new TreeSet<>(JpaMetamodelUtils.getEmbeddedPropertyNames((EmbeddableType<?>) ((SingularAttribute<?, ?>) lastAttribute).getType()));
+                    } else {
+                        embeddedPropertyNames = new TreeSet<>(cbf.getJpaProvider().getIdentifierOrUniqueKeyEmbeddedPropertyNames(cteType, attributeName));
+                    }
+
+                    PathExpression baseExpression = embeddedPropertyNames.size() > 1 ?
+                            ((PathExpression) selectExpression).clone(false) : ((PathExpression) selectExpression);
+
+                    for (String embeddedPropertyName : embeddedPropertyNames) {
+                        PathExpression pathExpression = firstBinding ?
+                                ((PathExpression) selectExpression) : baseExpression.clone(false);
+
+                        // TODO: this is supposed to check whether someAssociation is bound to someAlias.id
+                        // Actually, the id names can collide in the case of IdClass or EmbeddedIds, so a true check
+                        // should check whether the type of the path reference of the path expression is a basic type.
+                        // However, this seems not possible at this time because for some tests pathExpression.getPathReference()
+                        // returns null.
+                        if (embeddedPropertyName.equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
+                            throw new IllegalStateException("An association should be bound to its association type and not its identifier type");
```

Comment:
I'm throwing an error here now because as we discussed, we don't want users to bind entity types directly to identifier values.

Reply:
That's perfect!